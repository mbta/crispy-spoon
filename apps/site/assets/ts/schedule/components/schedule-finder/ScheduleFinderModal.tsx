import React, { ReactElement, useEffect, useState } from "react";
import { DirectionId, Route } from "../../../__v3api";
import {
  SimpleStopMap,
  RoutePatternsByDirection,
  ServiceInSelector,
  SelectedOrigin,
  ScheduleNote
} from "../__schedule";
import Modal from "../../../components/Modal";
import OriginModalContent from "./OriginModalContent";
import ScheduleModalContent from "./ScheduleModalContent";

export type Mode = "origin" | "schedule";

interface Props {
  closeModal: () => void;
  directionChanged?: (direction: DirectionId) => void;
  initialMode: Mode;
  initialDirection: DirectionId;
  initialOrigin: SelectedOrigin;
  originChanged?: (origin: SelectedOrigin) => void;
  route: Route;
  routePatternsByDirection: RoutePatternsByDirection;
  services: ServiceInSelector[];
  stops: SimpleStopMap;
  today: string;
  updateURL: (origin: SelectedOrigin, direction?: DirectionId) => void;
  handleOriginSelectClick: () => void;
  scheduleNote: ScheduleNote | null;
}

const ScheduleFinderModal = ({
  closeModal,
  directionChanged,
  initialMode,
  initialDirection,
  initialOrigin,
  originChanged,
  route,
  routePatternsByDirection,
  services,
  stops,
  today,
  updateURL,
  handleOriginSelectClick,
  scheduleNote
}: Props): ReactElement => {
  const [originState, setOriginState] = useState(initialOrigin);

  const getStopIdIfInDirection = (newDirection: DirectionId): string | null => {
    const directionDestination = route.direction_destinations[newDirection];
    const stopsForDirection = stops[newDirection];
    const stopsWithoutTerminal = stopsForDirection.filter(
      s => s.name !== directionDestination
    );
    // This needs to return null if the selected origin is the end of the line in the direction
    return stopsWithoutTerminal.find(s => s.id === originState)?.id || null;
  };

  const handleChangeDirection = (newDirection: DirectionId): void => {
    if (directionChanged) directionChanged(newDirection);

    // return the stop if it is present in the new direction
    // stops the user from having to reselect an origin on stop change
    const stop = getStopIdIfInDirection(newDirection);

    if (originChanged) originChanged(stop);
    updateURL(initialOrigin, newDirection);
  };

  const handleChangeOrigin = (newOrigin: SelectedOrigin): void => {
    setOriginState(newOrigin);
    if (originChanged) originChanged(newOrigin);
    updateURL(newOrigin, initialDirection);
  };

  const originModalContent = (): ReactElement => {
    const origin = initialOrigin;
    const direction = initialDirection;
    return (
      <OriginModalContent
        handleChangeOrigin={handleChangeOrigin}
        selectedOrigin={origin}
        stops={stops[direction] || []}
      />
    );
  };

  const scheduleModalContent = (scheduleOrigin: string): ReactElement => (
    <ScheduleModalContent
      handleChangeDirection={handleChangeDirection}
      handleChangeOrigin={handleChangeOrigin}
      handleOriginSelectClick={handleOriginSelectClick}
      route={route}
      routePatternsByDirection={routePatternsByDirection}
      selectedDirection={initialDirection}
      selectedOrigin={scheduleOrigin}
      services={services}
      stops={stops}
      today={today}
      scheduleNote={scheduleNote}
    />
  );

  const direction = initialDirection;
  const origin = initialOrigin;
  const originStop = stops[direction].find(stop => stop.id === origin);

  useEffect(() => {
    setOriginState(initialOrigin);
  }, [initialOrigin]);

  return (
    <Modal
      focusElementId={
        initialMode === "origin" ? "origin-filter" : "modal-close"
      }
      ariaLabel={{
        label:
          initialMode === "origin"
            ? "Choose Origin Stop"
            : `Schedules on the ${route.name} ${
                route.direction_names[direction]
              } to ${route.direction_destinations[direction]}${
                originStop ? ` from ${originStop.name}` : ""
              }`
      }}
      className={
        initialMode === "origin" ? "schedule-finder__origin-modal" : ""
      }
      closeModal={closeModal}
    >
      {initialMode === "origin" && originModalContent()}
      {initialMode === "schedule" &&
        origin !== null &&
        scheduleModalContent(origin)}
    </Modal>
  );
};

export default ScheduleFinderModal;
