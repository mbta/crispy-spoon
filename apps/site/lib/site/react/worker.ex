defmodule Site.React.Worker do
  @moduledoc """
  React renderer worker
  """
  require Logger
  use GenServer

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts)
  end

  def init(_opts) do
    node = System.find_executable("node") || System.find_executable("nodejs")
    config = Application.get_env(:site, :react)
    port = Port.open({:spawn_executable, node}, args: [config[:build_path]])
    {:ok, %{port: port}}
  end

  def handle_call({:render, name, props}, _from, %{port: port} = state) do
    body =
      Poison.encode!(%{
        name: name,
        props: props
      })

    Port.command(port, body <> "\n")

    response =
      ""
      |> receive_response(&handle_json/1)

    case response do
      %{data: %{"error" => nil} = data} ->
        {:reply, {:ok, data}, state}

      %{data: %{"error" => _error} = data} ->
        {:reply, {:error, data}, state}

      _ ->
        {:noreply, state}
    end
  end

  defp receive_response(prev_responses, handle_fn) do
    receive do
      {_, {:data, data}} ->
        str = to_string(data)
        resp = prev_responses <> str
        handle_response(resp, handle_fn)
    end
  end

  def handle_response(resp, handle_fn) do
    # Split and keep delimiter
    ~r/(?<=\n)/
    |> Regex.split(resp, trim: true)
    |> Enum.map(&handle_data(&1, handle_fn))
    |> Enum.find(&Map.has_key?(&1, :data))
  end

  def handle_data(str, handle_fn) do
    if String.ends_with?(str, "\n") do
      if String.starts_with?(str, "node_logging") do
        handle_logging(str)
      else
        handle_fn.(str)
      end
    else
      receive_response(str, handle_fn)
    end
  end

  def handle_json(msg) do
    json = Poison.decode!(msg)
    %{data: json}
  end

  def handle_logging(msg) do
    _ = Logger.warn(inspect(msg))
    %{}
  end

  def handle_info(msg, state) do
    _ = Logger.warn(inspect(msg))
    {:noreply, state}
  end
end
