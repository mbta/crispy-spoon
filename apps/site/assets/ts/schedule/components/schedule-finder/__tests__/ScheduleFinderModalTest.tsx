import React from "react";
import { mount } from "enzyme";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { DirectionId, Route } from "../../../../__v3api";
import {
  RoutePatternsByDirection,
  SelectedOrigin,
  ServiceInSelector
} from "../../__schedule";
import * as routePatternsByDirectionData from "../../__tests__/test-data/routePatternsByDirectionData.json";
import ScheduleFinderModal, { Mode as ModalMode } from "../ScheduleFinderModal";

jest.mock("../../../../helpers/use-fetch", () => ({
  __esModule: true,
  hasData: () => false,
  isLoading: () => true,
  isNotStarted: () => false,
  default: jest.fn().mockImplementation(() => [{ status: 2 }, jest.fn()])
}));

const routePatternsByDirection = routePatternsByDirectionData as RoutePatternsByDirection;

const services: ServiceInSelector[] = [
  {
    valid_days: [1, 2, 3, 4, 5],
    typicality: "typical_service",
    type: "weekday",
    start_date: "2019-07-08",
    removed_dates_notes: {},
    removed_dates: [],
    name: "Weekday",
    id: "BUS319-J-Wdy-02",
    end_date: "2019-08-30",
    description: "Weekday schedule",
    added_dates_notes: {},
    added_dates: [],
    rating_start_date: "2019-06-25",
    rating_end_date: "2019-10-25",
    rating_description: "Test",
    "default_service?": true
  },
  {
    valid_days: [6],
    typicality: "typical_service",
    type: "saturday",
    start_date: "2019-07-13",
    removed_dates_notes: {},
    removed_dates: [],
    name: "Saturday",
    id: "BUS319-K-Sa-02",
    end_date: "2019-08-31",
    description: "Saturday schedule",
    added_dates_notes: {},
    added_dates: [],
    rating_start_date: "2019-06-25",
    rating_end_date: "2019-10-25",
    rating_description: "Test",
    "default_service?": false
  },
  {
    valid_days: [7],
    typicality: "typical_service",
    type: "sunday",
    start_date: "2019-07-14",
    removed_dates_notes: {},
    removed_dates: [],
    name: "Sunday",
    id: "BUS319-L-Su-02",
    end_date: "2019-08-25",
    description: "Sunday schedule",
    added_dates_notes: {},
    added_dates: [],
    rating_start_date: "2019-06-25",
    rating_end_date: "2019-10-25",
    rating_description: "Test",
    "default_service?": false
  }
];

const today = "2019-12-05";

const route: Route = {
  description: "",
  direction_destinations: { 0: "Oak Grove", 1: "Forest Hills" },
  direction_names: { 0: "Inbound", 1: "Outbound" },
  id: "Orange",
  long_name: "Orange Line",
  name: "Orange",
  type: 1,
  line_id: null
};

const stops = {
  "1": [
    {
      name: "SL",
      id: "741",
      is_closed: false,
      zone: "1"
    },
    {
      name: "Abc",
      id: "123",
      is_closed: false,
      zone: null
    },
    {
      name: "Def",
      id: "456",
      is_closed: false,
      zone: null
    },
    {
      name: "Wellington",
      id: "place-welln",
      is_closed: true,
      zone: null
    }
  ],
  "0": [
    {
      name: "Wellington",
      id: "place-welln",
      is_closed: true,
      zone: null
    },
    {
      name: "Abc",
      id: "123",
      is_closed: false,
      zone: null
    },
    {
      name: "SL",
      id: "741",
      is_closed: false,
      zone: "1"
    }
  ]
};

describe("ScheduleFinderModal", () => {
  const mountComponent = (
    mode: ModalMode,
    direction: DirectionId,
    origin: SelectedOrigin,
    directionChanged?: (direction: DirectionId) => void,
    originChanged?: (origin: SelectedOrigin) => void
  ) =>
    mount(
      <ScheduleFinderModal
        closeModal={() => {}}
        initialMode={mode}
        route={route}
        routePatternsByDirection={routePatternsByDirection}
        services={services}
        stops={stops}
        today={today}
        updateURL={() => {}}
        initialDirection={direction}
        directionChanged={directionChanged}
        initialOrigin={origin}
        originChanged={originChanged}
        handleOriginSelectClick={() => {}}
        scheduleNote={null}
      />
    );

  it("matches snapshot in origin mode with origin selected", () => {
    const wrapper = mountComponent("origin", 0, null, undefined, undefined);
    expect(wrapper.debug()).toMatchSnapshot();
    wrapper.unmount();
  });

  it("matches snapshot in schedule mode", () => {
    const wrapper = mountComponent(
      "schedule",
      0,
      "place-welln",
      undefined,
      undefined
    );
    expect(wrapper.debug()).toMatchSnapshot();
    wrapper.unmount();
  });

  it("detects click and keyUp events in OriginListItem elements", () => {
    const originChangedSpy = jest.fn();

    const wrapper = mountComponent(
      "origin",
      0,
      null,
      undefined,
      originChangedSpy
    );

    // detect click event:
    wrapper
      .find(".schedule-finder__origin-list-item")
      .at(1)
      .simulate("click");

    expect(originChangedSpy).toHaveBeenCalledTimes(1);

    // detect keyUp event:
    originChangedSpy.mockRestore();

    wrapper
      .find(".schedule-finder__origin-list-item")
      .at(1)
      .simulate("keyUp", { key: "Enter" });

    expect(originChangedSpy).toHaveBeenCalledTimes(1);

    originChangedSpy.mockRestore();
    wrapper.unmount();
  });

  it("detects change in direction", () => {
    const spy = jest.fn();

    const wrapper = mountComponent(
      "schedule",
      0,
      "place-welln",
      jest.fn(),
      jest.fn()
    );

    wrapper.setProps({ updateURL: spy });

    // trigger change in direction:
    wrapper
      .find("select")
      .at(0)
      .simulate("change", { target: { value: "1" } });

    expect(spy).toHaveBeenCalledTimes(1);

    spy.mockRestore();
    wrapper.unmount();
  });

  it("searches for existing and non-existing stops", () => {
    const wrapper = mountComponent(
      "origin",
      0,
      "place-welln",
      jest.fn(),
      jest.fn()
    );

    // type on input field:
    wrapper
      .find("#origin-filter")
      .at(1) // select the input, not the SearchBox
      .simulate("change", { target: { value: "555 Street" } });

    expect(
      wrapper.find(".schedule-finder__origin-list").children().length
    ).toEqual(0);

    wrapper
      .find("#origin-filter")
      .at(1) // select the input, not the SearchBox
      .simulate("change", { target: { value: "Abc" } });

    expect(
      wrapper.find(".schedule-finder__origin-list").children().length
    ).toEqual(1);
    wrapper.unmount();
  });

  it("does not change origin if it exists on direction change", () => {
    const mode = "schedule";
    const direction = 0;
    const origin = "741";
    const directionChanged = jest.fn();
    const originChanged = jest.fn();
    const user = userEvent.setup();
    render(
      <ScheduleFinderModal
        closeModal={() => {}}
        initialMode={mode}
        route={route}
        routePatternsByDirection={routePatternsByDirection}
        services={services}
        stops={stops}
        today={today}
        updateURL={() => {}}
        initialDirection={direction}
        directionChanged={directionChanged}
        initialOrigin={origin}
        originChanged={originChanged}
        handleOriginSelectClick={() => {}}
        scheduleNote={null}
      />
    );

    // makes sure schedule finder is open
    expect(screen.getByText("Schedule Finder")).toBeInTheDocument();

    const directionSelect = screen.getByTestId(
      "schedule-finder-direction-select"
    );
    // couldn't get userEvent.selectOptions to work as expected
    fireEvent.click(directionSelect, { target: { value: 1 } });

    // The modal should not change to origin select
    expect(screen.getByText("Schedule Finder")).toBeInTheDocument();
    expect(
      (screen.getByRole("option", {
        name: "OUTBOUND Forest Hills"
      }) as HTMLOptionElement).selected
    ).toBe(true);
  });

  it("opens the select origin modal if the selected stop does not exist in the new direction", () => {
    const mode = "schedule";
    const direction = 0;
    const origin = "456";
    const directionChanged = jest.fn();
    const originChanged = jest.fn();
    const user = userEvent.setup();
    render(
      <ScheduleFinderModal
        closeModal={() => {}}
        initialMode={mode}
        route={route}
        routePatternsByDirection={routePatternsByDirection}
        services={services}
        stops={stops}
        today={today}
        updateURL={() => {}}
        initialDirection={direction}
        directionChanged={directionChanged}
        initialOrigin={origin}
        originChanged={originChanged}
        handleOriginSelectClick={() => {}}
        scheduleNote={null}
      />
    );

    // makes sure schedule finder is open
    expect(screen.getByText("Schedule Finder")).toBeInTheDocument();

    const directionSelect = screen.getByTestId(
      "schedule-finder-direction-select"
    );
    // couldn't get userEvent.selectOptions to work as expected
    fireEvent.click(directionSelect, { target: { value: 1 } });

    // The modal change to origin select
    expect(screen.getByText("Choose an origin stop")).toBeInTheDocument();
  });
});
