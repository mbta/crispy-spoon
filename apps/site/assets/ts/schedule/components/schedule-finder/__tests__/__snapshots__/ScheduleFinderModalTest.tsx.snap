// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScheduleFinderModal matches snapshot in origin mode 1`] = `
"<Component closeModal={[Function: closeModal]} initialMode=\\"origin\\" route={{...}} routePatternsByDirection={{...}} scheduleNote={{...}} services={{...}} stops={{...}} today=\\"2019-12-05\\" updateURL={[Function: updateURL]} initialDirection={0} directionChanged={[undefined]} initialOrigin={{...}} originChanged={[undefined]} handleOriginSelectClick={[Function: handleOriginSelectClick]}>
  <WrappedModal focusElementId=\\"origin-filter\\" ariaLabel={{...}} className=\\"schedule-finder__origin-modal\\" closeModal={[Function: closeModal]}>
    <Modal focusElementId=\\"origin-filter\\" ariaLabel={{...}} className=\\"schedule-finder__origin-modal\\" closeModal={[Function: closeModal]} priorPadding=\\"0px\\" paddingRight={0}>
      <ModalContent bodyPadding=\\"0px\\" scrollBarPadding={0} closeText=\\"Close\\" role=\\"dialog\\" ariaLabel={{...}} focusElementId=\\"origin-filter\\" className=\\"schedule-finder__origin-modal\\" closeModal={[Function: closeModal]}>
        <Portal containerInfo={{...}}>
          <aside aria-modal={true} aria-label=\\"Choose Origin Stop\\" role=\\"dialog\\" id=\\"modal-cover\\" className=\\"c-modal__cover\\" tabIndex={-1} onKeyDown={[Function: onKeyDown]} onClick={[Function: onClick]}>
            <div className=\\"c-modal schedule-finder__origin-modal\\">
              <button id=\\"modal-close\\" className=\\"btn btn-secondary c-modal__close\\" onClick={[Function: onClick]} type=\\"button\\">
                Close
              </button>
              <div className=\\"c-modal__content\\">
                <OriginModalContent handleChangeOrigin={[Function: handleChangeOrigin]} selectedOrigin={{...}} stops={{...}}>
                  <br />
                  <h3 className=\\"schedule-finder__origin-text\\">
                    Choose an origin stop
                  </h3>
                  <SearchBox id=\\"origin-filter\\" labelText=\\"Choose an origin stop\\" className=\\"schedule-finder__origin-search-container\\" placeholder=\\"Filter stops and stations\\" onChange={[Function: bound dispatchAction]}>
                    <div className=\\"schedule-finder__origin-search-container c-form__input-container\\">
                      <label htmlFor=\\"origin-filter\\" className=\\"sr-only\\">
                        Choose an origin stop
                      </label>
                      <input className=\\"c-form__input\\" id=\\"origin-filter\\" type=\\"text\\" onChange={[Function: onChange]} value=\\"\\" placeholder=\\"Filter stops and stations\\" />
                      <button className=\\"c-form__submit-btn\\" aria-label=\\"search\\" type=\\"button\\" onClick={[Function: onClick]}>
                        <span className=\\"notranslate c-svg__icon-search-reverse-default\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                      </button>
                    </div>
                  </SearchBox>
                  <p className=\\"schedule-finder__origin-text\\">
                    Select from the list below.
                  </p>
                  <div className=\\"schedule-finder__origin-list\\">
                    <OriginListItem stop={{...}} changeOrigin={[Function: handleChangeOrigin]} selectedOrigin={{...}} lastStop={{...}}>
                      <div tabIndex={0} role=\\"button\\" className=\\"schedule-finder__origin-list-item  disabled\\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                        <div className=\\"schedule-finder__origin-list-leftpad\\">
                           
                        </div>
                        Wellington
                         
                      </div>
                    </OriginListItem>
                    <OriginListItem stop={{...}} changeOrigin={[Function: handleChangeOrigin]} selectedOrigin={{...}} lastStop={{...}}>
                      <div tabIndex={0} role=\\"button\\" className=\\"schedule-finder__origin-list-item  \\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                        <div className=\\"schedule-finder__origin-list-leftpad\\">
                           
                        </div>
                        Abc
                         
                      </div>
                    </OriginListItem>
                    <OriginListItem stop={{...}} changeOrigin={[Function: handleChangeOrigin]} selectedOrigin={{...}} lastStop={{...}}>
                      <div tabIndex={0} role=\\"button\\" className=\\"schedule-finder__origin-list-item  disabled\\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                        <div className=\\"schedule-finder__origin-list-leftpad\\">
                           
                        </div>
                        SL
                         
                        <span className=\\"schedule-finder__zone\\">
                          Zone 
                          1
                        </span>
                      </div>
                    </OriginListItem>
                  </div>
                </OriginModalContent>
              </div>
            </div>
          </aside>
        </Portal>
      </ModalContent>
    </Modal>
  </WrappedModal>
</Component>"
`;

exports[`ScheduleFinderModal matches snapshot in origin mode with origin selected 1`] = `
"<Component closeModal={[Function: closeModal]} initialMode=\\"origin\\" route={{...}} routePatternsByDirection={{...}} scheduleNote={{...}} services={{...}} stops={{...}} today=\\"2019-12-05\\" updateURL={[Function: updateURL]} initialDirection={0} directionChanged={[undefined]} initialOrigin={{...}} originChanged={[undefined]} handleOriginSelectClick={[Function: handleOriginSelectClick]}>
  <WrappedModal focusElementId=\\"origin-filter\\" ariaLabel={{...}} className=\\"schedule-finder__origin-modal\\" closeModal={[Function: closeModal]}>
    <Modal focusElementId=\\"origin-filter\\" ariaLabel={{...}} className=\\"schedule-finder__origin-modal\\" closeModal={[Function: closeModal]} priorPadding=\\"0px\\" paddingRight={0}>
      <ModalContent bodyPadding=\\"0px\\" scrollBarPadding={0} closeText=\\"Close\\" role=\\"dialog\\" ariaLabel={{...}} focusElementId=\\"origin-filter\\" className=\\"schedule-finder__origin-modal\\" closeModal={[Function: closeModal]}>
        <Portal containerInfo={{...}}>
          <aside aria-modal={true} aria-label=\\"Choose Origin Stop\\" role=\\"dialog\\" id=\\"modal-cover\\" className=\\"c-modal__cover\\" tabIndex={-1} onKeyDown={[Function: onKeyDown]} onClick={[Function: onClick]}>
            <div className=\\"c-modal schedule-finder__origin-modal\\">
              <button id=\\"modal-close\\" className=\\"btn btn-secondary c-modal__close\\" onClick={[Function: onClick]} type=\\"button\\">
                Close
              </button>
              <div className=\\"c-modal__content\\">
                <OriginModalContent handleChangeOrigin={[Function: handleChangeOrigin]} selectedOrigin={{...}} stops={{...}}>
                  <br />
                  <h3 className=\\"schedule-finder__origin-text\\">
                    Choose an origin stop
                  </h3>
                  <SearchBox id=\\"origin-filter\\" labelText=\\"Choose an origin stop\\" className=\\"schedule-finder__origin-search-container\\" placeholder=\\"Filter stops and stations\\" onChange={[Function: bound dispatchAction]}>
                    <div className=\\"schedule-finder__origin-search-container c-form__input-container\\">
                      <label htmlFor=\\"origin-filter\\" className=\\"sr-only\\">
                        Choose an origin stop
                      </label>
                      <input className=\\"c-form__input\\" id=\\"origin-filter\\" type=\\"text\\" onChange={[Function: onChange]} value=\\"\\" placeholder=\\"Filter stops and stations\\" />
                      <button className=\\"c-form__submit-btn\\" aria-label=\\"search\\" type=\\"button\\" onClick={[Function: onClick]}>
                        <span className=\\"notranslate c-svg__icon-search-reverse-default\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                      </button>
                    </div>
                  </SearchBox>
                  <p className=\\"schedule-finder__origin-text\\">
                    Select from the list below.
                  </p>
                  <div className=\\"schedule-finder__origin-list\\">
                    <OriginListItem stop={{...}} changeOrigin={[Function: handleChangeOrigin]} selectedOrigin={{...}} lastStop={{...}}>
                      <div tabIndex={0} role=\\"button\\" className=\\"schedule-finder__origin-list-item  disabled\\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                        <div className=\\"schedule-finder__origin-list-leftpad\\">
                           
                        </div>
                        Wellington
                         
                      </div>
                    </OriginListItem>
                    <OriginListItem stop={{...}} changeOrigin={[Function: handleChangeOrigin]} selectedOrigin={{...}} lastStop={{...}}>
                      <div tabIndex={0} role=\\"button\\" className=\\"schedule-finder__origin-list-item  \\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                        <div className=\\"schedule-finder__origin-list-leftpad\\">
                           
                        </div>
                        Abc
                         
                      </div>
                    </OriginListItem>
                    <OriginListItem stop={{...}} changeOrigin={[Function: handleChangeOrigin]} selectedOrigin={{...}} lastStop={{...}}>
                      <div tabIndex={0} role=\\"button\\" className=\\"schedule-finder__origin-list-item  disabled\\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                        <div className=\\"schedule-finder__origin-list-leftpad\\">
                           
                        </div>
                        SL
                         
                        <span className=\\"schedule-finder__zone\\">
                          Zone 
                          1
                        </span>
                      </div>
                    </OriginListItem>
                  </div>
                </OriginModalContent>
              </div>
            </div>
          </aside>
        </Portal>
      </ModalContent>
    </Modal>
  </WrappedModal>
</Component>"
`;

exports[`ScheduleFinderModal matches snapshot in schedule mode 1`] = `
"<Component closeModal={[Function: closeModal]} initialMode=\\"schedule\\" route={{...}} routePatternsByDirection={{...}} scheduleNote={{...}} services={{...}} stops={{...}} today=\\"2019-12-05\\" updateURL={[Function: updateURL]} initialDirection={0} directionChanged={[undefined]} initialOrigin=\\"place-welln\\" originChanged={[undefined]} handleOriginSelectClick={[Function: handleOriginSelectClick]}>
  <WrappedModal focusElementId=\\"modal-close\\" ariaLabel={{...}} className=\\"\\" closeModal={[Function: closeModal]}>
    <Modal focusElementId=\\"modal-close\\" ariaLabel={{...}} className=\\"\\" closeModal={[Function: closeModal]} priorPadding=\\"0px\\" paddingRight={0}>
      <ModalContent bodyPadding=\\"0px\\" scrollBarPadding={0} closeText=\\"Close\\" role=\\"dialog\\" ariaLabel={{...}} focusElementId=\\"modal-close\\" className=\\"\\" closeModal={[Function: closeModal]}>
        <Portal containerInfo={{...}}>
          <aside aria-modal={true} aria-label=\\"Schedules to Inbound\\" role=\\"dialog\\" id=\\"modal-cover\\" className=\\"c-modal__cover\\" tabIndex={-1} onKeyDown={[Function: onKeyDown]} onClick={[Function: onClick]}>
            <div className=\\"c-modal \\">
              <button id=\\"modal-close\\" className=\\"btn btn-secondary c-modal__close\\" onClick={[Function: onClick]} type=\\"button\\">
                Close
              </button>
              <div className=\\"c-modal__content\\">
                <ScheduleModalContent handleChangeDirection={[Function: handleChangeDirection]} handleChangeOrigin={[Function: handleChangeOrigin]} handleOriginSelectClick={[Function: handleOriginSelectClick]} route={{...}} routePatternsByDirection={{...}} scheduleNote={{...}} selectedDirection={0} selectedOrigin=\\"place-welln\\" services={{...}} stops={{...}} today=\\"2019-12-05\\">
                  <div className=\\"schedule-finder schedule-finder--modal\\">
                    <Component onDirectionChange={[Function: handleChangeDirection]} onOriginChange={[Function: handleChangeOrigin]} onOriginSelectClick={[Function: handleOriginSelectClick]} route={{...}} selectedDirection={0} selectedOrigin=\\"place-welln\\" stopsByDirection={{...}}>
                      <form onSubmit={[Function: handleSubmit]}>
                        <h2 className=\\"h3 schedule-finder__heading\\">
                          <span className=\\"notranslate c-svg__icon\\" aria-hidden=\\"true\\" dangerouslySetInnerHTML={{...}} />
                           Schedule Finder
                        </h2>
                        <div className=\\"schedule-finder__prompt\\">
                          Choose a stop to get schedule information and real-time departure predictions.
                        </div>
                        <div className=\\"schedule-finder__inputs\\">
                          <label className=\\"schedule-finder__label\\">
                            Choose a direction
                            <SelectContainer>
                              <div tabIndex={0} className=\\"c-select-custom__container \\" role=\\"button\\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                                <select className=\\"c-select-custom\\" value={0} onChange={[Function: onChange]}>
                                  <option value={0}>
                                    INBOUND
                                     
                                    Oak Grove
                                  </option>
                                  <option value={1}>
                                    OUTBOUND
                                     
                                    Forest Hills
                                  </option>
                                </select>
                                <span className=\\"notranslate c-svg__icon c-select-custom__arrow\\" aria-hidden=\\"true\\" dangerouslySetInnerHTML={{...}} />
                              </div>
                            </SelectContainer>
                          </label>
                          <label className=\\"schedule-finder__label\\">
                            Choose an origin stop
                            <SelectContainer error={false} handleClick={[Function: handleOriginClick]}>
                              <div tabIndex={0} className=\\"c-select-custom__container \\" role=\\"button\\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                                <select className=\\"c-select-custom c-select-custom--noclick\\" value=\\"place-welln\\" onChange={[Function: onChange]}>
                                  <option value=\\"\\">
                                    Select
                                  </option>
                                  <option value=\\"place-welln\\">
                                    Wellington
                                  </option>
                                  <option value=\\"123\\">
                                    Abc
                                  </option>
                                  <option value=\\"741\\">
                                    SL
                                  </option>
                                </select>
                                <span className=\\"notranslate c-svg__icon c-select-custom__arrow\\" aria-hidden=\\"true\\" dangerouslySetInnerHTML={{...}} />
                              </div>
                            </SelectContainer>
                          </label>
                        </div>
                        <div className=\\"schedule-finder__submit text-right\\">
                          <input className=\\"btn btn-primary\\" type=\\"submit\\" value=\\"Get schedules\\" />
                        </div>
                      </form>
                    </Component>
                  </div>
                  <div className=\\"callout text-center u-bold\\">
                    There are no scheduled trips for 
                    December 5, 2019
                    .
                  </div>
                  <DailySchedule stopId=\\"place-welln\\" services={{...}} routeId=\\"Orange\\" directionId={0} routePatterns={{...}} today=\\"2019-12-05\\">
                    <h3>
                      Daily Schedule
                    </h3>
                    <SchedulesSelect sortedServices={{...}} todayServiceId=\\"\\" defaultSelectedServiceId=\\"BUS319-J-Wdy-02\\" todayDate={{...}} onSelectService={[Function: onSelectService]}>
                      <div className=\\"schedule-finder__service-selector\\">
                        <label htmlFor=\\"service_selector\\" className=\\"sr-only\\">
                          Schedules
                        </label>
                        <SelectContainer>
                          <div tabIndex={0} className=\\"c-select-custom__container \\" role=\\"button\\" onClick={[Function: onClick]} onKeyUp={[Function: onKeyUp]}>
                            <select id=\\"service_selector\\" className=\\"c-select-custom text-center u-bold\\" defaultValue=\\"BUS319-J-Wdy-02\\" onChange={[Function: onChange]}>
                              <ServiceOptGroup label=\\"Other Schedules\\" services={{...}} multipleWeekdays={false} todayServiceId=\\"\\">
                                <optgroup label=\\"Other Schedules\\">
                                  <option value=\\"BUS319-J-Wdy-02\\">
                                    Weekday schedule, Jul 8 to Aug 30
                                  </option>
                                  <option value=\\"BUS319-K-Sa-02\\">
                                    Saturday schedule, Jul 13 to Aug 31
                                  </option>
                                  <option value=\\"BUS319-L-Su-02\\">
                                    Sunday schedule, Jul 14 to Aug 25
                                  </option>
                                </optgroup>
                              </ServiceOptGroup>
                            </select>
                            <span className=\\"notranslate c-svg__icon c-select-custom__arrow\\" aria-hidden=\\"true\\" dangerouslySetInnerHTML={{...}} />
                          </div>
                        </SelectContainer>
                      </div>
                    </SchedulesSelect>
                    <Loading>
                      <div className=\\"c-spinner__container\\">
                        <div className=\\"c-spinner\\">
                          Loading...
                        </div>
                      </div>
                    </Loading>
                  </DailySchedule>
                </ScheduleModalContent>
              </div>
            </div>
          </aside>
        </Portal>
      </ModalContent>
    </Modal>
  </WrappedModal>
</Component>"
`;
